src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp

```diff
@@ -242,115 +303,130 @@ KOKKOS_INLINE_FUNCTION
-            ScalarT q = u_norm(cell,ipt) / ( u_norm(cell,ipt) + lambdaField(cell,ipt)*ice_softness(cell)*std::pow(Nval,n) );
-            beta(cell,ipt) = muCoulombField(cell,ipt) * Nval * std::pow( q, power) / u_norm(cell,ipt);
+  if(beta_type != GIVEN_CONSTANT) {
+    for (unsigned int ipt=0; ipt<dim; ++ipt) {
+
+      switch (effectivePressure_type) {
+      case EFFECTIVE_PRESSURE_TYPE::FIELD:
+        if(zero_N_on_floating_at_nodes) {
+          Nval = 0.0;
+          if(nodal) {
+            if (rho_i*thickness_field(cell,ipt)+rho_w*bed_topo_field(cell,ipt) > 0)
+              Nval = N(cell,ipt);
+          } else {
+            for (unsigned int node=0; node<numNodes; ++node)
+              if (rho_i*thickness_field(cell,node)+rho_w*bed_topo_field(cell,node) > 0)
+                Nval += N(cell,node)*BF(cell,node,ipt);
           }
-        } else {
-          for (unsigned int ipt=0; ipt<dim; ++ipt) {
-            ScalarT Nval = N(cell,ipt);
-            if (is_side_equation) Nval = KU::max(Nval, 0.0);
-            ScalarT q = u_norm(cell,ipt) / ( u_norm(cell,ipt) + lambdaField(cell,ipt)*ice_softness(cell)*std::pow(Nval,n) );
-            beta(cell,ipt) = mu * Nval * std::pow( q, power) / u_norm(cell,ipt);
+        } else
+          Nval = N(cell,ipt);
+        break;
+      case EFFECTIVE_PRESSURE_TYPE::ONE:
+        if(zero_N_on_floating_at_nodes) {
+          Nval = 0;
+          if(nodal) {
+            if(rho_i*thickness_field(cell,ipt)+rho_w*bed_topo_field(cell,ipt) > 0)
+              Nval =  1.0;
+          } else {
+            for (unsigned int node=0; node<numNodes; ++node)
+              if (rho_i*thickness_field(cell,node)+rho_w*bed_topo_field(cell,node) > 0)
+                Nval += BF(cell,node,ipt);
           }
         }
-      } else {
-        if (distributedMu) {
-          for (unsigned int ipt=0; ipt<dim; ++ipt) {
-            ScalarT Nval = N(cell,ipt);
-            if (is_side_equation) Nval = KU::max(Nval, 0.0);
-            ScalarT q = u_norm(cell,ipt) / ( u_norm(cell,ipt) + lambda*ice_softness(cell)*std::pow(Nval,n) );
-            beta(cell,ipt) = muCoulombField(cell,ipt) * Nval * std::pow( q, power) / u_norm(cell,ipt);
-          }
-        } else {
-          for (unsigned int ipt=0; ipt<dim; ++ipt) {
-            ScalarT Nval = N(cell,ipt);
-            if (is_side_equation) Nval = KU::max(Nval, 0.0);
-            ScalarT q = u_norm(cell,ipt) / ( u_norm(cell,ipt) + lambda*ice_softness(cell)*std::pow(Nval,n) );
-            beta(cell,ipt) = mu * Nval * std::pow( q, power) / u_norm(cell,ipt);
+        break;
+      case EFFECTIVE_PRESSURE_TYPE::HYDROSTATIC:
+        if(nodal)
+          Nval = g*KU::max(rho_i*thickness_field(cell,ipt)+KU::min(rho_w*bed_topo_field(cell,ipt),0.0),0.0);
+        else {
+          MeshScalarT thickness(0), bed_topo(0);
+          for (unsigned int node=0; node<numNodes; ++node) {
+            thickness += thickness_field(cell,node)*BF(cell,node,ipt);
```

Comment:
We almost surely already compute these fields at the quadrature points, so why not simply asking for thickness and bed topo as QuadPoints MDField's?

Reply:
To avoid storing both thickness_field and thickness_field_at_nodes (and same for bed_topo), given that in some cases you would need both.